src/backend/sgx/attestation.rs

```diff
@@ -0,0 +1,317 @@
+
+        let response = Message::parse_from_bytes(&res_bytes)?;
+
+        Ok(response)
+    }
+}
+
+fn get_key_id_num() -> Result<u32, Error> {
+    let mut req = AesmRequest::new();
+
+    let mut msg = Request_GetSupportedAttKeyIDNumRequest::new();
+    msg.set_timeout(AESM_REQUEST_TIMEOUT);
+    req.set_getSupportedAttKeyIDNumReq(msg);
+
+    let pb_msg = req.send()?;
+
+    let res = pb_msg.get_getSupportedAttKeyIDNumRes();
+
+    if res.get_errorCode() != 0 {
+        return Err(Error::new(
+            ErrorKind::Other,
+            format!(
+                "Received error code {:?} in GetSupportedAttKeyIDNum",
+                res.get_errorCode()
+            ),
+        ));
+    }
+
+    Ok(res.get_att_key_id_num())
+}
+
+fn get_key_ids(num_key_ids: u32) -> Result<Vec<Vec<u8>>, Error> {
+    let expected_buffer_size: u32 = num_key_ids * SGX_KEY_ID_SIZE;
+
+    let mut req = AesmRequest::new();
+
+    let mut msg = Request_GetSupportedAttKeyIDsRequest::new();
+    msg.set_timeout(AESM_REQUEST_TIMEOUT);
+    msg.set_buf_size(expected_buffer_size);
+    req.set_getSupportedAttKeyIDsReq(msg);
+
+    let pb_msg = req.send()?;
+
+    let res = pb_msg.get_getSupportedAttKeyIDsRes();
+
+    if res.get_errorCode() != 0 {
+        return Err(Error::new(
+            ErrorKind::InvalidData,
+            format!("GetSupportedAttKeyIDs: error: {:?}", res.get_errorCode()),
+        ));
+    }
+
+    let key_ids_blob = res.get_att_key_ids();
+    Ok(key_ids_blob
+        .chunks_exact(SGX_KEY_ID_SIZE as usize)
+        .map(Vec::from)
+        .collect())
+}
+
+/// Gets Att Key ID
+fn get_attestation_key_id() -> Result<Vec<u8>, Error> {
+    let num_key_ids = get_key_id_num()?;
+    if num_key_ids != 1 {
+        return Err(Error::new(
+            ErrorKind::Other,
+            format!("Unexpected number of key IDs: {} != 1", num_key_ids),
+        ));
+    }
+
+    let key_ids = get_key_ids(num_key_ids)?;
+
+    if key_ids.len() != 1 {
+        return Err(Error::new(
+            ErrorKind::Other,
+            format!(
+                "GeSupportedAttKeyIDs: invalid count: {} != 1",
+                key_ids.len()
+            ),
+        ));
+    }
+
+    Ok(key_ids.get(0).unwrap().clone())
+}
+
+/// Fills the Target Info of the QE into the output buffer specified and
+/// returns the number of bytes written.
+fn get_target_info(akid: Vec<u8>, size: usize, out_buf: &mut [u8]) -> Result<usize, Error> {
+    if out_buf.len() != SGX_TI_SIZE {
+        return Err(Error::new(
+            ErrorKind::Other,
+            format!(
+                "Invalid output buffer size: {} != {}",
+                out_buf.len(),
+                SGX_TI_SIZE
+            ),
+        ));
+    }
+
+    let mut req = AesmRequest::new();
+    let mut msg = Request_InitQuoteExRequest::new();
+
+    msg.set_timeout(AESM_REQUEST_TIMEOUT);
+    msg.set_b_pub_key_id(true);
+    msg.set_att_key_id(akid);
+    msg.set_buf_size(size as u64);
```

Comment:
nit; compiler should be able to infer the type here
```suggestion
    msg.set_buf_size(size as _);
```